test/model/src/include/model/driver/kv_workload_runner.h

```diff
@@ -0,0 +1,325 @@
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef MODEL_DRIVER_KV_WORKLOAD_RUNNER_H
+#define MODEL_DRIVER_KV_WORKLOAD_RUNNER_H
+
+#include <shared_mutex>
+#include <unordered_map>
+#include "model/driver/kv_workload.h"
+#include "model/core.h"
+#include "model/kv_database.h"
+#include "model/kv_table.h"
+#include "model/kv_transaction.h"
+#include "wiredtiger.h"
+
+namespace model {
+
+/*
+ * kv_workload_runner --
+ *     The workload runner for the model.
+ */
+class kv_workload_runner {
+
+public:
+    /*
+     * kv_workload_runner::kv_workload_runner --
+     *     Create a new workload runner.
+     */
+    inline kv_workload_runner(kv_database &database) : _database(database) {}
+
+    /*
+     * kv_workload_runner::database --
+     *     Get the database.
+     */
+    inline kv_database &
+    database() const noexcept
+    {
+        return _database;
+    }
+
+    /*
+     * kv_workload_runner::run_operation --
+     *     Run the given operation.
+     */
+    inline void
+    run_operation(const operation::any &op)
+    {
+        std::visit(
+          [this](auto &&x) {
+              int ret = do_operation(x);
+              /*
+               * In the future, we would like to be able to test operations that can fail, at which
+               * point we would record and compare return codes. But we're not there yet, so just
+               * fail on error.
+               */
+              if (ret != 0)
+                  throw wiredtiger_exception(ret);
+          },
+          op);
+    }
+
+protected:
+    /*
+     * kv_workload_runner::do_operation --
+     *     Execute the given workload operation in the model.
+     */
+    int
+    do_operation(const operation::begin_transaction &op)
+    {
+        add_transaction(op.txn_id, _database.begin_transaction());
+        return 0;
+    }
+
+    /*
+     * kv_workload_runner::do_operation --
+     *     Execute the given workload operation in the model.
+     */
+    int
+    do_operation(const operation::checkpoint &op)
+    {
+        _database.create_checkpoint(op.name == "" ? nullptr : op.name.c_str());
+        return 0;
+    }
+
+    /*
+     * kv_workload_runner::do_operation --
+     *     Execute the given workload operation in the model.
+     */
+    int
+    do_operation(const operation::commit_transaction &op)
+    {
+        remove_transaction(op.txn_id)->commit(op.commit_timestamp, op.durable_timestamp);
```

Comment:
It looks a bit unnatural to call `remove_transaction` to perform the `commit` operation.
Should we `commit` and then `remove` the transaction?

Reply:
This is just a short way of ensuring that all transactions in the collection are uncommitted, so we first remove a transaction from the map and then commit it. (This works because we use shared pointers.) I added a comment explaining this.

If it still a bad style, I can change it... I don't really expect a problem with transactions being accessed incorrectly from different threads.

(And I guess this is also a habit from many years of Java programming...)

Reply:
If you remove the transaction from the map but don't do anything about it. Then try to start a transaction, would that fail?

Reply:
(Responding just now, as I just got back from PTO.) If you remove the transaction from the map, that should not affect the ability to start any other transactions. The transaction ID used here is meaningful only within the scope of the workload specification, so as long as the workload does something to remove the transaction from the map (whether by commit, rollback, or database restart), it can reuse the same ID immediately for a new transaction. The ID is just a way to specify which operations belong to which transactions in the workload specification and has no special meaning to WT or the model.

Reply:
But for WT, we should not have more than one transaction for a given session. If this is detected by the model, we should raise an error saying something went wrong, either in the model or what the model is parsing.

Reply:
There is no concept of sessions at the workload level; the workload executor for WT takes care of mapping the workload operations to WT operations, including session and cursor management. There would never be a case where two transactions would be mapped to one session (that is, unless I have a bug somewhere in the executor).